iris-mpc-cpu/src/protocol/ops.rs

```diff
@@ -757,6 +758,161 @@ pub async fn min_of_pair_batch(
+    session: &mut Session,
+    distances: &[DistanceShare<u32>],
+    batch_size: usize,
+) -> Result<Vec<DistanceShare<u32>>> {
+    if distances.is_empty() {
+        eyre::bail!("Expected at least one distance share");
+    }
+    if distances.len() % batch_size != 0 {
+        eyre::bail!("Distances length must be a multiple of batch size");
+    }
+
+    // Within each batch, compute all the pairwise comparisons in a round-robin fashion.
+    // The resulting comparison table looks like
+    //    | d0 | d1 | d2 | d3 |
+    // ------------------------
+    // d0 | -  | b01| b02| b03|
+    // d1 |    | -  | b12| b13|
+    // d2 |    |    | -  | b23|
+    // d3 |    |    |    | -  |
+    // where bij is the bit corresponding to di < dj.
+    // Comparison bits are arranged in a flat vector as
+    // [b01, b02, b03, b12, b13, b23]
+    let num_batches = distances.len() / batch_size;
+    let mut pairs = Vec::with_capacity(num_batches * (batch_size * (batch_size - 1) / 2));
+    for i_batch in 0..num_batches {
+        for i in 0..batch_size {
+            for j in (i + 1)..batch_size {
+                let distance_i = distances[i * num_batches + i_batch].clone();
+                let distance_j = distances[j * num_batches + i_batch].clone();
+                pairs.push((distance_i, distance_j));
+            }
+        }
+    }
+    let comparison_bits = oblivious_cross_compare(session, &pairs).await?;
+    // Fill in the rest of the comparison table by setting diagonal bits to 1 and negating the bits above the diagonal.
+    //    | d0 | d1 | d2 | d3 |
+    // ------------------------
+    // d0 | 1  | b01| b02| b03|
+    // d1 |!b01| 1  | b12| b13|
+    // d2 |!b02|!b12| 1  | b23|
+    // d3 |!b03|!b13|!b23| 1  |
+    // Extract this table column-wise to AND them element-wise.
+    // Group ith columns together, i.e., return a matrix M, where M[i] contains for the comparison bits between distance i of every batch and all the other distances within the same batch.
+    let mut batch_selection_bits = (0..batch_size)
+        .map(|_| Vec::with_capacity(num_batches * batch_size))
+        .collect_vec();
+    for batch in comparison_bits.chunks(batch_size * (batch_size - 1) / 2) {
+        let mut batch_matrix: Vec<Vec<Share<Bit>>> = (0..batch_size)
+            .map(|_| Vec::with_capacity(batch_size))
+            .collect();
+        let mut batch_counter = 0;
+        for i in 0..batch_size {
+            let row = &batch[batch_counter..batch_counter + (batch_size - i - 1)];
+            // fill in the ith column
+            batch_matrix[i].push(Share::from_const(Bit::new(true), session.own_role()));
+            for bit in row {
+                batch_matrix[i].push(bit.not());
+            }
+            // fill the other columns
+            for j in (i + 1)..batch_size {
+                batch_matrix[j].push(row[j - i - 1].clone());
+            }
+            batch_counter += batch_size - i - 1;
+        }
```

Comment:
I find the suggested change below easier to read.
```suggestion
         for i in 0..batch_size {
            for j in 0..batch_size {
                let value = match i.cmp(&j) {
                    Ordering::Less => {
                        batch_counter += 1;
                        batch[batch_counter - 1].clone()
                    }
                    Ordering::Equal => Share::from_const(Bit::new(true), session.own_role()),
                    Ordering::Greater => batch_matrix[i][j].not(),
                };
                batch_matrix[j].push(value);
            }
        }
```